lighthouse-core/lib/page-functions.js

```diff
@@ -6,8 +6,39 @@
 // @ts-nocheck
 'use strict';
 
+/** @typedef {HTMLElementTagNameMap & {[id: string]: HTMLElement}} HTMLElementByTagName */
+
 /* global window document Node ShadowRoot */
 
+/**
+ * Creates valid JavaScript code given functions, strings of valid code, and arguments.
+ * @template T, R
+ * @param {(...args: T[]) => R} mainFn The main function to call. It's return value will be the return value
+ * of `createEvalCode`, wrapped in a Promise.
+ * @param {{mode?: 'iffe'|'function', args?: T[], deps?: Function[]}} _ Set mode to `iffe` to
+ * create a self-executing function expression, set to `function` to create just a function
+ * declaration statement. Args should match the args of `mainFn`, and can be any serializable
+ * value. `deps` are functions that must be defined for `mainFn` to work.
+ */
+function createEvalCode(mainFn, {mode, args, deps} = {}) {
```

Comment:
could this be directly incorporated into `evaluate()` (and `evaluateFunctionOnObject`) instead of splitting it into this file? It's mostly not shared code since the two callers take different branches in the conditional (and it's not really a "page function" in the same way as the other functions in this file are).

Reply:
I prefer it here bc/ it's easier to test (driver has some overhead with `beforeEach` etc.) and 
keeps the driver file smaller (it's so big). Maybe `lib/eval.js`?

And then export two function: `createFunctionEvalCode` and `createIffeEvalCode`?

Reply:
> I prefer it here bc/ it's easier to test (driver has some overhead with `beforeEach` etc.) and
> keeps the driver file smaller (it's so big). Maybe `lib/eval.js`?
> 
> And then export two function: `createFunctionEvalCode` and `createIffeEvalCode`?

well it would definitely be good to split the two, but it'll also have a new home after #11633, and it also looks trivial to add to `.evaluateAsync` in `driver-test`? 

```
const {expression} = connectionStub.sendCommand.findInvocation('Runtime.evaluate');
expect(expression).toEqual(`(() => {
function mainFn() {
  return true;
}
return mainFn();
})()`);
```

Reply:
@patrickhulce would you prefer this going into `driver.js` or a new file `lib/eval.js`?

Reply:
new `lib/eval.js` file SGTM, though I'll also +1 @brendankenny's test suggest to make sure `.evaluate` uses it :)

Reply:
> new lib/eval.js file

Isn't the (currently named) `RuntimeController` going to be this already in #11633, though? I guess I don't understand what the problem is just inlining the five lines or so :)

Reply:
> Isn't the (currently named) RuntimeController going to be this already in #11633, though?

Ya, that's why I don't really care much where it lives in this PR :) 

> I guess I don't understand what the problem is just inlining the five lines or so

I mean I don't feel terribly strongly about it, the rebase gets slightly more annoying with inline but not by much.